rfd/0174-kube-rbac-bootstrap.md

```diff
@@ -0,0 +1,158 @@
+the default Kubernetes user-facing roles and adding the ability to define RBAC setup to be
+automatically provisioned to Kubernetes clusters.
+
+## Why
+
+Currently, users have to rely on third-party tools or manual actions
+when making sure their Kubernetes clusters have the correct roles/role bindings and that those are in sync with the RBAC setup defined in Teleport.
+The changes we propose will make it easier for clients to start using Teleport for accessing Kubernetes clusters.
+
+## Scope
+
+This RFD is focused only on Kubernetes resources related to RBAC functionality, other types of resources are out of scope,
+though they might be added in the future.
+
+## Details
+
+### Exposing the default Kubernetes user facing cluster roles (cluster-admin, view, edit)
+
+Kubernetes clusters have default user-facing cluster roles: cluster-admin, view, edit. By default they are not usable, because
+they are not exposed through role bindings to any subjects (the cluster-admin role has internal Kubernetes bindings). 
+Currently, we only expose the cluster-admin role on GKE
+installations, since system:masters group is not available for impersonation there. We will expand on this and when
+discovering a cluster/installing kube-agent/creating a service account, we will
+create cluster role bindings for those roles, accordingly linking them to the groups "default-cluster-admin", "default-view" and
+"default-edit". It will give users an opportunity to always have standard set of Kubernetes permissions which they can use together
+with Teleport's fine-grained RBAC definitions for Kubernetes. This will make enrolling and starting using Kubernetes clusters
+into Teleport a much simpler process, since user will be able to setup RBAC completely in Teleport if they want. For example,
+role giving read-only access to staging namespace pods might look like this:
+
+```yaml
+kind: role
+metadata:
+  name: staging-kube-access
+version: v7
+spec:
+  allow:
+    kubernetes_labels:
+      'region': '*'
+      'platform': 'aws'
+    kubernetes_resources:
+      - kind: pod
+        namespace: "staging"
+        name: "*"
+    kubernetes_groups:
+    - default-view
+    kubernetes_users:
+    - developer
+  deny: {}
+```
+
+Users will not need to create a separate role binding in Kubernetes for it to work - once the cluster is enrolled in Teleport, the default-view is already there.
+We will add an option to the teleport-kube-agent chart and to our
+kubeconfig generation script to not create those cluster role bindings, but it will be enabled by default. 
+
+### Automatically provisioning RBAC resources to Kubernetes clusters.
+
+We will add a new type of resources, called "KubeProvision", where user will be able to specify Kubernetes RBAC resources they want to
+be provisioned onto their Kubernetes clusters.
+clusters. A single Teleport resource of that type can define multiple Kubernetes RBAC resources. There's a limit on size of a resource Teleport
```

Comment:
for simplicity, can we split the roles, clusterrole, rolebinding and cluster role binding objects?

Reply:
We don't need to create 4 separate resources but the resource struct must separate those fields like:
```
metadata:
   name: ...
 spec:
     clusterRoleBinding:
         ----
     clusterRole:
 ----
      roleBinding:
         ----
     role:

```
So we have a simple way of understanding what's a role and rolebinding objects...
Inside each one we have the required fields

Reply:
we can transform them into arrays if needed

Reply:
Actually I think it's not that bad. Having a max one of clusterRole, roleBinding, ... allow us to map 1:1 between those resources and the teleport resource name for easier debug